File: leetcode/lc-0013/solution.cpp
class Solution {
public:
    int romanToInt(string s) {
        auto toInt = [](char c) {
            switch (c) {
            case 'I': return 1;
            case 'V': return 5;
            case 'X': return 10;
            case 'L': return 50;
            case 'C': return 100;
            case 'D': return 500;
            case 'M': return 1000;
            default: return 0;
            }
        };
        int num = 0;
        int prev = 0;
        int n = s.size();
        for (int i = 0; i < n; i++) {
            int curr = toInt(s[i]);
            if (curr > prev) {
                num += curr - 2 * prev;
            } else {
                num += curr;
            }
            prev = curr;
        }
        return num;
    }
};

File: leetcode/lc-0013/solution.rs
impl Solution {
    pub fn roman_to_int(s: String) -> i32 {
        let to_int = |c| {
            match c {
            b'I' => 1,
            b'V' => 5,
            b'X' => 10,
            b'L' => 50,
            b'C' => 100,
            b'D' => 500,
            b'M' => 1000,
            _ => panic!("unreachable"),
            }
        };
        let mut num = 0;
        let mut prev = 0;
        let n = s.len();
        for i in 0..n {
            let curr = to_int(s.as_bytes()[i]);
            if (curr > prev) {
                num += curr - 2 * prev;
            } else {
                num += curr;
            }
            prev = curr;
        }
        return num;
    }
}
